File: Combinatorial Game Theory/Alpha-Beta Pruning/alphabeta.cpp
// C++ program to demonstrate working of Alpha-Beta Pruning
#include <bits/stdc++.h>

using namespace std;

const int MAX = 1000;
const int MIN = -1000;

int minimax(int depth, int nodeIndex, bool maximizingPlayer, int values[], int alpha, int beta)
{

  if (depth == 3)
    return values[nodeIndex];

  if (maximizingPlayer)
  {
    int best = MIN;

    for (int i = 0; i < 2; i++)
    {
      int val = minimax(depth + 1, nodeIndex * 2 + i, false, values, alpha, beta);
      best = max(best, val);
      alpha = max(alpha, best);

      if (beta <= alpha)
        break;
    }
    return best;
  }
  else
  {
    int best = MAX;
    for (int i = 0; i < 2; i++)
    {
      int val = minimax(depth + 1, nodeIndex * 2 + i, true, values, alpha, beta);
      best = min(best, val);
      beta = min(beta, best);
      if (beta <= alpha)
        break;
    }
    return best;
  }
}

int main()
{
  int values[8] = {3, 5, 6, 9, 1, 2, 0, -1};
  cout << "The optimal value is : " << minimax(0, 0, true, values, MIN, MAX);
  return 0;
}

File: Combinatorial Game Theory/Alpha-Beta Pruning/alphabeta.py
#!/usr/bin/python3

# ------------------------------------------------------#
#       alphabeta(input, start, upper, lower)           #
#           - input: list                               #
#           - start: int (either 0 or 1)                #
#           - upper: float (max value in input list)    #
#           - lower: float (min value in input list)    #
# ------------------------------------------------------#

example = [[[5, 1, 2], [8, -8, -9]], [[9, 4, 5], [-3, 4, 3]]]
root = 0
pruned = 0

def subtree(branch, depth, alpha, beta):
    global tree
    global root
    global pruned
    i = 0
    for child in branch:
        if type(child) is list:
            (nalpha, nbeta) = subtree(child, depth + 1, alpha, beta)
            if depth % 2 == 1:
                beta = nalpha if nalpha < beta else beta
            else:
                alpha = nbeta if nbeta > alpha else alpha
            branch[i] = alpha if depth % 2 == 0 else beta
            i += 1
        else:
            if depth % 2 == 0 and alpha < child:
                alpha = child
            if depth % 2 == 1 and beta > child:
                beta = child
            if alpha >= beta:
                pruned += 1
                break
    if depth == root:
        tree = alpha if root == 0 else beta
    return (alpha, beta)

def alphabeta(in_tree=tree, start=root, upper=-15, lower=15):
    global tree
    global pruned
    global root

    (alpha, beta) = subtree(tree, start, upper, lower)
    
    if __name__ == "__main__":
        print (f'(alpha, beta)  {alpha}, {beta}')
        print (f'result  {tree}')
        print (f'tree pruned {pruned} times.')

    return (alpha, beta, tree, pruned)

if __name__ == "__main__":
    alphabeta(None)